### PR TITLE
SAM3 and SAM3.1 ONNX and TensorRT Export

### LINKED ISSUE #23023: Is there any way to export the TensorRT engine file for SAM3?
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I know that most models can export TensorRT engine files simply by calling the method `export()`. I’m wondering whether SAM 3 also supports exporting to TensorRT.

### Additional

_No response_

### PR DESCRIPTION
**Pull request by onuralpszr**
Fix: https://github.com/ultralytics/ultralytics/issues/23023

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Adds end-to-end ONNX and TensorRT export and inference support for SAM3 and SAM3.1, including text, box, and point-prompt segmentation workflows.

### 📊 Key Changes
- Adds SAM3 ONNX export wrappers for vision, text, prompt, mask, and decoder components, plus TensorRT engine conversion.
- Introduces a multi-file `SAM3Backend` for ONNX Runtime and TensorRT inference with fixed image-size handling and optional point-prompt modules.
- Adds TensorRT-compatible attention, RoPE, GridSample-based ROI alignment, GELU, and dynamic-shape export adaptations.
- Supports SAM3.1 checkpoint key renaming and reports missing or randomly initialized weights during loading.
- Extends `SAM` prediction and export APIs to support exported SAM3 directories, text prompts, and point prompts.
- Adds API reference documentation and navigation entries for the new backend and export modules.

### 🎯 Purpose & Impact
- Enables deployment of SAM3 and SAM3.1 models as ONNX or TensorRT artifacts for production and edge inference.
- Improves TensorRT compatibility by avoiding unavailable ROIAlign plugins and addressing FP16 attention precision issues.
- Provides a unified inference interface for semantic, box-prompt, and point-prompt segmentation.
- Adds safeguards and warnings for incompatible or incomplete checkpoints, helping prevent silently incorrect outputs.
- Expands documentation coverage for the new export and backend functionality.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @onuralpszr, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms receipt of your SAM3 and SAM3.1 ONNX/TensorRT export contribution; an Ultralytics engineer will assist with the review. Please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit changes to the **minimum** necessary for this feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The PR adds a comprehensive modular SAM3 export/backend path, but exported inference currently has blocking device and fixed-image-size integration issues. GPU ONNX selection, predictor stride handling, and non-default-size point exports need correction before this is reliable end to end.

💬 Posted 3 inline comments

File: ultralytics/models/sam/predict.py [hunk 1]
```diff
@@ -2219,6 +2219,20 @@ class SAM3SemanticPredictor(SAM3Predictor):
...
+        model_path = Path(self.args.model)
+        # Support ONNX/TensorRT directory backends
+        if model_path.is_dir() and (model_path.name.endswith("_onnx") or model_path.name.endswith("_engine")):
+            from ultralytics.nn.backends.sam3 import SAM3Backend
+
+            backend = SAM3Backend(model_path, device=self.args.device or "cpu")
+            # The graphs are traced at a fixed size, so adopt it instead of letting the caller's
+            # default reach a graph that can only reject it.
+            if backend.imgsz:
+                self.args.imgsz = backend.imgsz
```
Comment (UltralyticsAssistant) [inline ultralytics/models/sam/predict.py:2243]: ⚠️ **HIGH**: Adopting the exported size here is undone during predictor setup: inherited `Predictor.setup_model()` unconditionally sets `self.model.stride = 32`, and `setup_source()` subsequently runs `check_imgsz()` using that stride. The default export size, 1008, is not divisible by 32, so it is rounded to 1024 before inference and the fixed-size vision graph rejects the input. Preserve the SAM3 backend stride of 14 (or bypass stride rounding for exported SAM3 graphs) before source setup.

File: ultralytics/nn/backends/sam3.py [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+                f"Pass imgsz={self.imgsz}, or re-export at {want}."
+            )
+
+    def eval(self):
+        """Return self, the exported graphs are always in inference mode."""
+        return self
+
+    def to(self, device):
+        """Move subsequent outputs to ``device`` and return self."""
+        self.device = torch.device(device) if isinstance(device, str) else device
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/backends/sam3.py]: ⚠️ **HIGH**: `SAM(..._onnx).predict(device="cuda")` still executes all ONNX sessions on CPU. The backend is constructed during `SAM._load()` with the default CPU device, so `_load_onnx()` selects `CPUExecutionProvider`; later predictor setup calls this `to()` method, which only changes `self.device` and never recreates the sessions/providers. This makes the documented CUDA inference path silently run the expensive graphs on CPU (with extra CPU↔GPU copies). Reinitialize the ONNX sessions with CUDA providers when moving devices, or defer backend construction until the predictor-selected device is known.

File: ultralytics/utils/export/sam3_onnx.py [hunk 3]
```diff
@@ -0,0 +1,1147 @@
...
+    device = torch.device(device) if isinstance(device, str) else device
+    assert imgsz % 14 == 0, f"imgsz={imgsz} must be divisible by patch_size=14"
+
+    LOGGER.info(f"\n{prefix} building SAM3SemanticModel from {checkpoint_path}...")
+    model = build_sam3_image_model(checkpoint_path, enable_segmentation=True)
+    model = model.to(device).eval()
+    for p in model.parameters():
+        p.requires_grad = False
+
+    if imgsz != 1008:
```
Comment (UltralyticsAssistant) [inline ultralytics/utils/export/sam3_onnx.py:808]: ⚠️ **HIGH**: Non-1008 exports with point modules cannot complete. This resizes only the semantic vision model, so `fpn2` becomes `imgsz / 14` spatially, but `tracker_model` is always built with `image_size=1008`; its prompt encoder still emits `dense_embeddings` and `dense_pe` at 72×72. The later mask-decoder export combines those fixed 72×72 tensors with the resized SAM2-neck feature map, causing a spatial-shape mismatch for any supported `imgsz` other than 1008. Either configure the interactive model/prompt encoder for the requested image size or reject non-1008 point-prompt exports.

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-08-11 08:24:25 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/31472973818): `ci.yml` run 29095
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/31472975245): `docker.yml` run 2956

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-08-25 06:49:13 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/32818570331): `ci.yml` run 30340
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/32818571951): `docker.yml` run 3057

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-08-25 16:05:03 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/32869681923): `ci.yml` run 30453
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/32869684629): `docker.yml` run 3062

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-08-25 18:53:14 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/32886274918): `ci.yml` run 30482
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/32886277192): `docker.yml` run 3065
